go/tools/gazelle/resolve/index.go

```diff
@@ -0,0 +1,358 @@
+	r, err := ix.findRuleByImport(imp, lang, fromRel)
+	if err != nil {
+		return NoLabel, err
+	}
+	return r.label, nil
+}
+
+func getGoImportPath(r bf.Rule, goPrefix, buildRel string) string {
+	// TODO(#597): account for subdirectory where goPrefix was set, when we
+	// support multiple prefixes.
+	imp := r.AttrString("importpath")
+	if imp != "" {
+		return imp
+	}
+	imp = path.Join(goPrefix, buildRel)
+	if name := r.Name(); name != config.DefaultLibName {
+		imp = path.Join(imp, name)
+	}
+	return imp
+}
+
+func findGoProtoSources(ix *RuleIndex, r *ruleRecord) []importSpec {
+	rule := bf.Rule{Call: r.call}
+	protoExpr, ok := rule.Attr("proto").(*bf.StringExpr)
+	if !ok {
+		return nil
+	}
+	protoLabel, err := ParseLabel(protoExpr.Value)
+	if err != nil {
+		return nil
+	}
+	protoRule, err := ix.findRuleByLabel(protoLabel, r.label.Pkg)
+	if err != nil {
+		return nil
+	}
+	var importedAs []importSpec
+	for _, source := range findSources(bf.Rule{Call: protoRule.call}, protoRule.label.Pkg, ".proto") {
+		importedAs = append(importedAs, importSpec{lang: config.ProtoLang, imp: source})
+	}
+	return importedAs
+}
+
+func findSources(r bf.Rule, buildRel, ext string) []string {
+	srcsExpr := r.Attr("srcs")
+	srcsList, ok := srcsExpr.(*bf.ListExpr)
+	if !ok {
+		return nil
+	}
+	var srcs []string
+	for _, srcExpr := range srcsList.List {
+		src, ok := srcExpr.(*bf.StringExpr)
+		if !ok {
+			continue
+		}
+		label, err := ParseLabel(src.Value)
+		if err != nil || !label.Relative || !strings.HasSuffix(label.Name, ext) {
+			continue
+		}
+		srcs = append(srcs, path.Join(buildRel, label.Name))
+	}
+	return srcs
+}
+
+func findDefaultVisibility(oldFile *bf.File, buildRel string) visibilitySpec {
+	if oldFile == nil {
+		return visibilitySpec{config.PrivateVisibility}
+	}
+	for _, stmt := range oldFile.Stmt {
+		call, ok := stmt.(*bf.CallExpr)
+		if !ok {
+			continue
+		}
+		rule := bf.Rule{Call: call}
+		if rule.Kind() == "package" {
+			return parseVisibility(rule.Attr("default_visibility"), buildRel)
+		}
+	}
+	return visibilitySpec{config.PrivateVisibility}
+}
+
+func parseVisibility(visExpr bf.Expr, buildRel string) visibilitySpec {
+	visList, ok := visExpr.(*bf.ListExpr)
+	if !ok {
+		return visibilitySpec{config.PrivateVisibility}
+	}
+	var visibility visibilitySpec
+	for _, elemExpr := range visList.List {
+		elemStr, ok := elemExpr.(*bf.StringExpr)
+		if !ok {
+			continue
+		}
+		if elemStr.Value == config.PublicVisibility || elemStr.Value == config.PrivateVisibility {
+			visibility = append(visibility, elemStr.Value)
+			continue
+		}
+		label, err := ParseLabel(elemStr.Value)
+		if err != nil {
+			continue
+		}
+		label = label.Abs("", buildRel)
+		if label.Repo != "" || label.Name != "__subpackages__" {
+			continue
+		}
+		visibility = append(visibility, label.Pkg)
+	}
+	return visibility
+}
+
+func isVisibleFrom(visibility visibilitySpec, defRel, useRel string) bool {
```

Comment:
I am not convinced we should be handling visibility, I think it is going to be more surprising than helpful, especially as it's going to be hard to be sure we cover all the visibility cases.
I think the set of cases where visibility is going to help us pick the right one of multiple packages with the same import path are negligible to non existent, far more likely are the cases where we should generate the dep anyway so they get a nice "is not visible" error when they try to build the thing they really wanted.

Reply:
You're right, I think I half-realized this when I allowed findRuleByImport to succeed when the only matches available were restricted by visibility.

I've removed all visibility parsing.